# Learned cost-pressure frontier (λ = 0 / 20 / 80)

One Colab session. Same Drive repo, 80k Tevatron-NQ slice, and Qwen setup as `FreeCost_Trainer_Sanity_CA_RL_ARAG.ipynb`.

**Paper headline:** train **one** tiny policy per cost-pressure level. Put all of them on the **EM vs dollars** plane next to the three frozen dots (naive / rule / max_tools). The spread **is** the figure.

The 2026-09-13 λ=80 / `act_penalty` 0–0.02 family sat entirely on the −EV side (extra max_tools $ × 80 ≈ 0.046 vs ~0.028 quality). Every policy correctly collapsed to retrieve→stop. This sweep **crosses break-even**:

| Preset | lambda_cost | act_penalty | $ term on extra tools | Expected story |
|---|---:|---:|---:|---|
| `frontier_lambda0` | 0 | 0.0 | 0 | Free end — tools clearly win (`correctness_only` already used them here) |
| `frontier_lambda20` | 20 | 0.0 | ≈ 0.012 | Still +EV vs the 0.028 quality gain |
| `frontier_act02` | 80 | 0.02 | ≈ 0.046 (+ action tax) | Expensive end — spend should lose |

Quality weights stay at `default`. **40 epochs** per preset (not 5). Each epoch logs sampled `mean_reward` **and** greedy `eval_reward` (argmax on the same 100 train questions). Frozen exams also use argmax; without the greedy curve we cannot tell reward vs training apart.

Does **not** overwrite `learned_policy.pt` or the ranking plots. Frozen naive/rule/max dots are read from the existing `*_default.json` files (no re-eval). Frontier exams use `--no-figures` and write `pilot_summary_frontier_lambda0.json` / `_lambda20.json` / `_act02.json`.

`--policies learned` **merges** into an existing `pilot_summary_<preset>.json`. It does not drop rule / max. The canonical **default** ranking table is four policies. Housekeeping cells at the end rebuild that table if needed.

Skip the `git reset --hard` cell the first time if this notebook is not on origin yet.

In [7]:
import os
from google.colab import userdata

# 1. Mount Google Drive to save files permanently
from google.colab import drive
drive.mount('/content/drive')

# # 2. Retrieve your secure token
# TOKEN = userdata.get('GITHUB_TOKEN')
# USERNAME = "Smartcobra"
# REPO_NAME = "ca-rl-arag"

# # 3. Configure Git credentials
# !git config --global user.name "jitu"
# !git config --global user.email "jitu.learn@gmail.com"

# # 4. Clone the repository into Google Drive
# %cd /content/drive/MyDrive/carlarag
# !git clone https://{USERNAME}:{TOKEN}@github.com/{USERNAME}/{REPO_NAME}.git


Mounted at /content/drive


In [8]:
#goto repo directory
%cd /content/drive/MyDrive/carlarag/ca-rl-arag/

/content/drive/MyDrive/carlarag/ca-rl-arag


In [9]:
# Load HF_TOKEN from Colab secrets. Do not write tokens into the notebook or a committed .env.
from google.colab import userdata
import os

os.environ["HF_TOKEN"] = userdata.get("HF_TOKEN")


In [12]:
#git pull command
# Revert/reset local changes
!git fetch origin
!git checkout feature_baseline_v2
!git reset --hard origin/feature_baseline_v2
!git clean -fd

# Pull latest changes
!git pull origin feature_baseline_v2

remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 15 (delta 12), reused 15 (delta 12), pack-reused 0 (from 0)
Unpacking objects: 100% (15/15), 8.74 KiB | 5.00 KiB/s, done.
From https://github.com/Smartcobra/ca-rl-arag
   9be9c19..cab3b9c  feature_baseline_v2 -> origin/feature_baseline_v2
Already on 'feature_baseline_v2'
Your branch is behind 'origin/feature_baseline_v2' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
HEAD is now at cab3b9c cost-pressure frontier (lambda=80, act_penalty sweep)
From https://github.com/Smartcobra/ca-rl-arag
 * branch            feature_baseline_v2 -> FETCH_HEAD
Already up to date.


In [13]:
# ##Git Pull
!git log -5
!git pull


commit cab3b9cc8a38fc36c29c108403a4f66a4886a1b8 (HEAD -> feature_baseline_v2, origin/feature_baseline_v2)
Author: parsuram <parsuram.pradhan@yobytech.com>
Date:   Sun Sep 13 19:59:37 2026 +0530

    cost-pressure frontier (lambda=80, act_penalty sweep)

commit 9be9c1905c9552efe9a6921a51f4aa84b7d285db
Author: parsuram <parsuram.pradhan@yobytech.com>
Date:   Sun Sep 13 19:36:05 2026 +0530

    correctness_only vs lambda_zero- updated .md

commit 6dbfe49a450dc0844eca3f90eca6f6941417c633
Author: jitu <jitu.learn@gmail.com>
Date:   Sun Sep 13 06:35:13 2026 +0000

    Colab-Execution: free-cost trainer sanity (correctness_only vs lambda_zero)

commit 81092e63459e1cc2dc0e7251a2cc65b7dd6ce775
Author: parsuram <parsuram.pradhan@yobytech.com>
Date:   Sun Sep 13 09:50:31 2026 +0530

    correctness_only vs lambda_zero

commit efdd9abb8add2112ca4cfff55f7ea12a198057c3
Author: parsuram <parsuram.pradhan@yobytech.com>
Date:   Thu Sep 10 01:13:10 2026 +0530

    updated notes and md and added learned 

In [14]:
# Install dependencies
!pip install -r requirements.txt

INFO: pip is looking at multiple versions of multiprocess to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.3/527.3 kB 16.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.6/177.6 kB 18.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 146.7/146.7 kB 16.6 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.12.0
    Uninstalling fsspec-2025.12.0:
      Successfully uninstalled fsspec-2025.12.0
  Attempting uninstall: dill
    Found existing installation: dill 0.4.1
    Uninstalling dill-0.4.1:
      Successfully uninstalled dill-0.4.1
  Attempting uninstall: multiprocess
    Found existing installation: multiprocess 0.70.19
    Uninstalling multiprocess-0.70.19:
      Succe

## Data (same locked 80k slice)

`smoke_test.py` overwrites `data/processed/`. Always run `prepare_data.py --hf` after it.

If Drive already has the ranking slice (`n_eval=300`, `n_passages=80000`, `n_nq_anchor=0`), skip the next five cells and jump to **Train 1**.

In [15]:
#run Smoke
!python scripts/smoke_test.py

SMOKE OK
baseline_pred: Cairo em= 1.0 reward= 1.5247
agent_pred: Cairo actions= ['retrieve', 'rerank', 'verify', 'stop'] em= 1.0


In [16]:
!python scripts/prepare_data.py --hf

Loading HotpotQA (distractor) from HuggingFace...
Generating train split: 100% 90447/90447 [00:06<00:00, 14472.53 examples/s]
Generating validation split: 100% 7405/7405 [00:00<00:00, 32670.21 examples/s]
Trying single-hop source: DPR Wikipedia 100-word passages (Tevatron/wikipedia-nq)
  loading Tevatron/wikipedia-nq split=train (train, n=40)...
  loading Tevatron/wikipedia-nq split=test (eval, n=150)...
  loading Tevatron/wikipedia-nq split=dev (eval, n=150)...
Single-hop source OK: DPR Wikipedia 100-word passages (Tevatron/wikipedia-nq) (train=40 eval=150 wiki_passages=2959 eval_gold_articles=847)
Adding unused Hotpot distractors toward 80000 passages (slice currently 5045; NQ uses DPR Wikipedia 100-word passages (Tevatron/wikipedia-nq))...
  scanned 2000 unused Hotpot rows...
  scanned 4000 unused Hotpot rows...
  scanned 6000 unused Hotpot rows...
  scanned 8000 unused Hotpot rows...
Distractor pool: +74955 passages from 8491 unused rows (corpus=80000, target=80000).
Computing BM25

In [17]:
##First check: printed / saved meta
import json
from pathlib import Path

p = Path("data/processed/slice_meta.json")
meta = json.loads(p.read_text())
print(json.dumps({
    "source": meta["source"],
    "n_eval": meta["n_eval"],
    "eval_by_dataset": meta["eval_by_dataset"],
    "n_passages": meta["n_passages"],
    "n_nq_anchor": meta["n_nq_anchor"],
    "nq_corpus": (meta.get("distractor_pool") or {}).get("single_hop", {}).get("nq_corpus"),
}, indent=2))
assert meta["n_eval"] == 300, meta["n_eval"]
assert meta["n_passages"] >= 50000, meta["n_passages"]
assert meta["n_nq_anchor"] == 0, meta["n_nq_anchor"]

{
  "source": "huggingface_nq_hotpot",
  "n_eval": 300,
  "eval_by_dataset": {
    "hotpot_qa": 150,
    "natural_questions": 150
  },
  "n_passages": 80000,
  "n_nq_anchor": 0,
  "nq_corpus": "dpr_wikipedia_w100"
}


In [18]:
#Confirm questions did not grow
from collections import Counter
from src.utils import read_jsonl

eval_set = read_jsonl("data/processed/eval_slice.jsonl")
train = read_jsonl("data/processed/train_slice.jsonl")
print(len(train), Counter(e["dataset"] for e in train))
print(len(eval_set), Counter(e["dataset"] for e in eval_set))

100 Counter({'hotpot_qa': 60, 'natural_questions': 40})
300 Counter({'hotpot_qa': 150, 'natural_questions': 150})


In [19]:
#Confirm golds stayed attached, pool did not leak into examples
from src.data.loaders import HOTPOT_DISTRACTOR_SOURCE
from src.utils import read_jsonl

corpus = read_jsonl("data/processed/corpus.jsonl")
eval_set = read_jsonl("data/processed/eval_slice.jsonl")

pool_ids = {p["passage_id"] for p in corpus if p.get("source") == HOTPOT_DISTRACTOR_SOURCE}
local = {pid for ex in eval_set + read_jsonl("data/processed/train_slice.jsonl")
         for pid in ex.get("local_passage_ids") or []}

print("pool passages", len(pool_ids))
print("pool ids leaked into example local_passage_ids", len(pool_ids & local))  # want 0

hotpot = [e for e in eval_set if e["dataset"] == "hotpot_qa"]
missing = sum(1 for e in hotpot if not e.get("local_passage_ids") or not e.get("supporting_titles"))
print("hotpot eval missing golds", missing)  # want 0

pool passages 74955
pool ids leaked into example local_passage_ids 0
hotpot eval missing golds 0


In [20]:
#The check that matters: recall dropped on Hotpot
!python scripts/retrieval_diagnostics.py

{
  "overall": {
    "recall@1": 0.52,
    "recall@5": 0.7566666666666667,
    "recall@20": 0.9666666666666667,
    "n_examples": 300,
    "n_miss_at_5": 73,
    "n_corpus": 80000
  },
  "by_dataset": {
    "hotpot_qa": {
      "recall@1": 0.76,
      "recall@5": 0.9266666666666666,
      "recall@20": 0.9866666666666667,
      "n_examples": 150,
      "n_miss_at_5": 11,
      "n_corpus": 80000
    },
    "natural_questions": {
      "recall@1": 0.28,
      "recall@5": 0.5866666666666667,
      "recall@20": 0.9466666666666667,
      "n_examples": 150,
      "n_miss_at_5": 62,
      "n_corpus": 80000
    }
  }
}


## Train + eval (do not skip)

Each train is **40 epochs** × 100 train questions, plus a **greedy (argmax) pass** on those same 100 after every epoch (`eval_reward` / `eval_n_steps` / `eval_n_verify` in the curve JSON). Each exam is frozen **argmax** on the 300.

`--epochs 40` is passed explicitly so a stale yaml cannot silently train 5 epochs. `_best.pt` is the best **greedy** `eval_reward`, not the sampled train mean.

`--policies learned` scores only the new checkpoint and **merges** that row into `pilot_summary_<preset>.json`. It does not drop rule / max. These frontier exams also pass `--no-figures` so ranking plots stay the `default` table.

Watch **both** sampled and greedy columns. Sampled `mean_verify` with greedy `eval_verify=0` is the gap that previously erased tool use. Two greedy steps = retrieve then stop.

Expect ~10× the 2026-09-13 wall clock (40 epochs + a greedy train pass each epoch). Budget a couple of days on a T4 for all three pairs. After the sweep, the housekeeping cells rebuild `pilot_summary_default.json` (four policies) if that file is missing rows.

In [21]:
# Train 1/3 — frontier_lambda0 (lambda=0, P_act=0). Free end.
!python scripts/train_policy.py \
  --reward-preset frontier_lambda0 \
  --epochs 40 \
  --checkpoint results/checkpoints/learned_policy_frontier_lambda0.pt \
  --curve results/metrics/train_policy_curve_frontier_lambda0.json

Ranking data check OK: train=100 {'hotpot_qa': 60, 'natural_questions': 40} corpus=80000 (>= 50000)
TRAIN ONLY path=/content/drive/MyDrive/carlarag/ca-rl-arag/data/processed/train_slice.jsonl n=100 by_dataset={'hotpot_qa': 60, 'natural_questions': 40} preset=frontier_act0 hidden=16 epochs=5 lr=0.003
[GPU] before model creation: name=Tesla T4 total=14.56 GiB allocated=0.00 GiB reserved=0.00 GiB free=14.46 GiB
config.json: 100% 661/661 [00:00<00:00, 2.57MB/s]
tokenizer_config.json: 100% 7.30k/7.30k [00:00<00:00, 20.2MB/s]
vocab.json: 100% 2.78M/2.78M [00:00<00:00, 74.1MB/s]
merges.txt: 100% 1.67M/1.67M [00:00<00:00, 94.5MB/s]
tokenizer.json: 100% 7.03M/7.03M [00:00<00:00, 146MB/s]
model.safetensors.index.json: 100% 35.6k/35.6k [00:00<00:00, 85.1MB/s]
Reconstructing (incomplete total...): |          |  0.00B /  0.00B            

Fetching 2 files:   0% 0/2 [00:00<?, ?it/s]
Reconstructing (incomplete total...):   0% 0.00/6.17G [00:00<?, ?B/s]         
Reconstructing (incomplete total...): 

In [22]:
# Exam 1/3 — freeze frontier_lambda0 on the 300
!python scripts/run_pilot.py \
  --reward-preset frontier_lambda0 \
  --policies learned \
  --learned-checkpoint results/checkpoints/learned_policy_frontier_lambda0.pt \
  --no-figures

Ranking data check OK: eval=300 {'hotpot_qa': 150, 'natural_questions': 150} corpus=80000 (>= 50000)
Corpus=80000 examples=300 by_dataset={'hotpot_qa': 150, 'natural_questions': 150} preset=frontier_act0
[GPU] before model creation: name=Tesla T4 total=14.56 GiB allocated=0.00 GiB reserved=0.00 GiB free=14.46 GiB
Loading weights: 100% 434/434 [00:24<00:00, 17.41it/s]
[GPU] after model creation: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
[GPU] before learned: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
agent:learned: 100% 300/300 [06:03<00:00,  1.21s/it]
learned overall: EM=0.333 F1=0.397 reward=0.587 abstain=0.150 n_correct=100/300 n_abstained=45 steps=2.00 retrieve=1.00 verify=0.00
  HotpotQA: EM=0.393 F1=0.445 reward=0.639 abstain=0.193 n_correct=59/150 n_abstained=29 steps=2.00 retrieve=1.00 verify=0.00
  Natural Questions: EM=0.273 F1=0.348 reward=0.535 abstain=0.107 n_correct=41/150 n_abstained=16 steps=2.0

In [23]:
# Train 2/3 — frontier_lambda20 (lambda=20, P_act=0). Still +EV on paper.
!python scripts/train_policy.py \
  --reward-preset frontier_lambda20 \
  --epochs 40 \
  --checkpoint results/checkpoints/learned_policy_frontier_lambda20.pt \
  --curve results/metrics/train_policy_curve_frontier_lambda20.json

Ranking data check OK: train=100 {'hotpot_qa': 60, 'natural_questions': 40} corpus=80000 (>= 50000)
TRAIN ONLY path=/content/drive/MyDrive/carlarag/ca-rl-arag/data/processed/train_slice.jsonl n=100 by_dataset={'hotpot_qa': 60, 'natural_questions': 40} preset=frontier_act005 hidden=16 epochs=5 lr=0.003
[GPU] before model creation: name=Tesla T4 total=14.56 GiB allocated=0.00 GiB reserved=0.00 GiB free=14.46 GiB
Loading weights: 100% 434/434 [00:24<00:00, 17.70it/s]
[GPU] after model creation: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
epoch 1/5  n=100  mean_reward=0.6740  mean_em=0.420  mean_steps=4.75  mean_retrieve=1.62  mean_verify=0.46
epoch 2/5  n=100  mean_reward=0.6779  mean_em=0.420  mean_steps=3.97  mean_retrieve=1.36  mean_verify=0.37
epoch 3/5  n=100  mean_reward=0.6975  mean_em=0.440  mean_steps=3.36  mean_retrieve=1.27  mean_verify=0.27
epoch 4/5  n=100  mean_reward=0.6457  mean_em=0.390  mean_steps=3.34  mean_retrieve=1.23  mean_verify

In [24]:
# Exam 2/3 — freeze frontier_lambda20 on the 300
!python scripts/run_pilot.py \
  --reward-preset frontier_lambda20 \
  --policies learned \
  --learned-checkpoint results/checkpoints/learned_policy_frontier_lambda20.pt \
  --no-figures

Ranking data check OK: eval=300 {'hotpot_qa': 150, 'natural_questions': 150} corpus=80000 (>= 50000)
Corpus=80000 examples=300 by_dataset={'hotpot_qa': 150, 'natural_questions': 150} preset=frontier_act005
[GPU] before model creation: name=Tesla T4 total=14.56 GiB allocated=0.00 GiB reserved=0.00 GiB free=14.46 GiB
Loading weights: 100% 434/434 [00:25<00:00, 17.06it/s]
[GPU] after model creation: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
[GPU] before learned: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
agent:learned: 100% 300/300 [06:11<00:00,  1.24s/it]
learned overall: EM=0.333 F1=0.397 reward=0.582 abstain=0.150 n_correct=100/300 n_abstained=45 steps=2.00 retrieve=1.00 verify=0.00
  HotpotQA: EM=0.393 F1=0.445 reward=0.634 abstain=0.193 n_correct=59/150 n_abstained=29 steps=2.00 retrieve=1.00 verify=0.00
  Natural Questions: EM=0.273 F1=0.348 reward=0.530 abstain=0.107 n_correct=41/150 n_abstained=16 steps=2

In [25]:
# Train 3/3 — frontier_act02 (lambda=80, P_act=0.02). Expensive end.
!python scripts/train_policy.py \
  --reward-preset frontier_act02 \
  --epochs 40 \
  --checkpoint results/checkpoints/learned_policy_frontier_act02.pt \
  --curve results/metrics/train_policy_curve_frontier_act02.json

Ranking data check OK: train=100 {'hotpot_qa': 60, 'natural_questions': 40} corpus=80000 (>= 50000)
TRAIN ONLY path=/content/drive/MyDrive/carlarag/ca-rl-arag/data/processed/train_slice.jsonl n=100 by_dataset={'hotpot_qa': 60, 'natural_questions': 40} preset=frontier_act02 hidden=16 epochs=5 lr=0.003
[GPU] before model creation: name=Tesla T4 total=14.56 GiB allocated=0.00 GiB reserved=0.00 GiB free=14.46 GiB
Loading weights: 100% 434/434 [00:25<00:00, 17.02it/s]
[GPU] after model creation: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
epoch 1/5  n=100  mean_reward=0.5700  mean_em=0.390  mean_steps=4.61  mean_retrieve=1.59  mean_verify=0.47
epoch 2/5  n=100  mean_reward=0.6169  mean_em=0.400  mean_steps=4.25  mean_retrieve=1.69  mean_verify=0.35
epoch 3/5  n=100  mean_reward=0.6077  mean_em=0.400  mean_steps=3.68  mean_retrieve=1.62  mean_verify=0.28
epoch 4/5  n=100  mean_reward=0.6333  mean_em=0.400  mean_steps=3.05  mean_retrieve=1.30  mean_verify=

In [26]:
# Exam 3/3 — freeze frontier_act02 on the 300
!python scripts/run_pilot.py \
  --reward-preset frontier_act02 \
  --policies learned \
  --learned-checkpoint results/checkpoints/learned_policy_frontier_act02.pt \
  --no-figures

Ranking data check OK: eval=300 {'hotpot_qa': 150, 'natural_questions': 150} corpus=80000 (>= 50000)
Corpus=80000 examples=300 by_dataset={'hotpot_qa': 150, 'natural_questions': 150} preset=frontier_act02
[GPU] before model creation: name=Tesla T4 total=14.56 GiB allocated=0.00 GiB reserved=0.00 GiB free=14.46 GiB
Loading weights: 100% 434/434 [00:24<00:00, 17.53it/s]
[GPU] after model creation: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
[GPU] before learned: name=Tesla T4 total=14.56 GiB allocated=5.75 GiB reserved=5.76 GiB free=8.70 GiB
agent:learned: 100% 300/300 [06:12<00:00,  1.24s/it]
learned overall: EM=0.333 F1=0.397 reward=0.567 abstain=0.150 n_correct=100/300 n_abstained=45 steps=2.00 retrieve=1.00 verify=0.00
  HotpotQA: EM=0.393 F1=0.445 reward=0.619 abstain=0.193 n_correct=59/150 n_abstained=29 steps=2.00 retrieve=1.00 verify=0.00
  Natural Questions: EM=0.273 F1=0.348 reward=0.515 abstain=0.107 n_correct=41/150 n_abstained=16 steps=2.

## How to check the results

Run the next cell after any exam finishes. It writes `results/metrics/frontier_sweep_table.json` and `results/figs/frontier_em_usd.png`.

### Files

| What | λ=0 | λ=20 | λ=80 act=0.02 |
|---|---|---|---|
| Train curve | `train_policy_curve_frontier_lambda0.json` | `..._lambda20.json` | `..._act02.json` |
| Checkpoint | `learned_policy_frontier_lambda0.pt` | `..._lambda20.pt` | `..._act02.pt` |
| Eval | `learned_frontier_lambda0.json` | `..._lambda20.json` | `..._act02.json` |

Each curve row has sampled `mean_*` **and** greedy `eval_reward` / `eval_n_steps` / `eval_n_verify` (`eval_split: train_greedy`). That greedy column is the same decision rule as the 300-exam.

Frozen dots (already on disk): `baseline_default.json`, `rule_based_default.json`, `max_tools_default.json`.

### What to look at

The figure is **EM (y) vs mean $ (x)**. The three learned policies should land at **different** points.

| You see | Meaning |
|---|---|
| λ=0 uses tools (greedy steps > 2, verify > 0) | Free end works; trainer + greedy eval agree |
| λ=20 spends some, EM between naive and max_tools | Mid point is on the +EV side |
| λ=80 act=0.02 sits on naive (2 steps, EM 0.333) | Expensive end, as designed |
| All three learned dots stack on naive **and** greedy `eval_verify` stayed 0 | Reward is still −EV, or 40 epochs were not enough — the curve tells you which |
| Sampled verify > 0 but greedy `eval_verify` = 0 | Same entropy/mode trap as 2026-09-10; do not cite sampled reward |

In [27]:
# Read frontier artifacts and draw EM vs $. Safe after any subset finishes.
import json
import sys
from pathlib import Path

sys.path.insert(0, "scripts")
from plot_results import collect_frontier_table, plot_frontier_em_usd

from src.utils import read_jsonl

metrics = Path("results/metrics")
figs = Path("results/figs")
table = collect_frontier_table(metrics)
table_path = metrics / "frontier_sweep_table.json"
table_path.write_text(json.dumps(table, indent=2), encoding="utf-8")
print(f"Wrote {table_path}")

print("\nFrozen ranking anchors (default reward, 300-eval)")
print(f"{'policy':<18} {'EM':>6} {'$':>10} {'steps':>7} {'retr':>6} {'ver':>6} {'correct':>8}")
for name, row in (table.get("frozen") or {}).items():
    print(
        f"{name:<18} {row['mean_em']:6.3f} {row['mean_total_usd']:10.2e} "
        f"{row['mean_n_steps']:7.2f} {row['mean_n_retrieve']:6.2f} {row['mean_n_verify']:6.2f} "
        f"{int(row['n_correct'])}/{int(row['n_examples'])}"
    )

print("\nLearned family (λ=0 / 20 / 80, locked 300-eval)")
for preset, row in (table.get("learned") or {}).items():
    print(
        f"{preset:<18} {row['mean_em']:6.3f} {row['mean_total_usd']:10.2e} "
        f"{row['mean_n_steps']:7.2f} {row['mean_n_retrieve']:6.2f} {row['mean_n_verify']:6.2f} "
        f"{int(row['n_correct'])}/{int(row['n_examples'])}"
    )
    curve_p = metrics / f"train_policy_curve_{preset}.json"
    if curve_p.exists():
        curve = json.loads(curve_p.read_text())
        print(
            "  sampled: "
            + ", ".join(
                f"ep{r.get('epoch')}:steps={r.get('mean_n_steps'):.2f}/ver={r.get('mean_n_verify'):.2f}"
                for r in curve
            )
        )
        print(
            "  greedy:  "
            + ", ".join(
                f"ep{r.get('epoch')}:eval_r={float(r.get('eval_reward') or 0):.3f}"
                f"/steps={float(r.get('eval_n_steps') or 0):.2f}"
                f"/ver={float(r.get('eval_n_verify') or 0):.2f}"
                for r in curve
            )
        )
    traj = Path(f"results/trajectories/learned_{preset}.jsonl")
    if traj.exists():
        rows = read_jsonl(traj)
        n_two = sum(
            1 for r in rows
            if float(r.get("n_steps") or 0) <= 2
            and float(r.get("n_retrieve") or 0) <= 1
            and float(r.get("n_verify") or 0) <= 0
        )
        print(f"  retrieve->stop rows: {n_two}/{len(rows)}")

if table.get("learned"):
    plot_frontier_em_usd(table, figs)
    print("Open results/figs/frontier_em_usd.png")
else:
    print("No learned_frontier_*.json yet — run a train+exam pair first.")


Wrote results/metrics/frontier_sweep_table.json

Frozen ranking anchors (default reward, 300-eval)
policy             EM          $   steps   retr    ver  correct
naive_rag       0.333   1.72e-04    2.00   1.00   0.00 100/300
rule_based      0.323   5.04e-04    4.00   1.00   1.00 97/300
max_tools       0.350   7.47e-04    7.00   3.00   1.00 105/300

Learned family (lambda=80, locked 300-eval)
frontier_act0   0.333   1.72e-04    2.00   1.00   0.00 100/300
  train epochs: ep1:steps=4.75/ver=0.46, ep2:steps=3.97/ver=0.36, ep3:steps=3.36/ver=0.29, ep4:steps=3.02/ver=0.14, ep5:steps=2.93/ver=0.14
  retrieve->stop rows: 300/300
frontier_act005  0.333   1.72e-04    2.00   1.00   0.00 100/300
  train epochs: ep1:steps=4.75/ver=0.46, ep2:steps=3.97/ver=0.37, ep3:steps=3.36/ver=0.27, ep4:steps=3.34/ver=0.19, ep5:steps=3.50/ver=0.37
  retrieve->stop rows: 300/300
frontier_act02  0.333   1.72e-04    2.00   1.00   0.00 100/300
  train epochs: ep1:steps=4.61/ver=0.47, ep2:steps=4.25/ver=0.35, ep3:st

## Housekeeping: four-policy ranking summary

The frontier sweep above writes `pilot_summary_frontier_lambda0.json` / `_lambda20.json` / `_act02.json`. It does **not** rebuild `pilot_summary_default.json`.

`--policies learned` **merges** into the existing summary. It does not drop rule / max.

The canonical **default** table is naive / rule / max / learned.

| Step | Command | GPU? | When |
|---|---|---|---|
| Merge test | `python tests/test_pilot_summary_merge.py` | no | always safe |
| Learned-only exam | `python scripts/run_pilot.py --policies learned` | yes | `.pt` exists; replace only the learned row |
| Four-policy rebuild | `python scripts/run_pilot.py --run-env-check` | yes | reward or slice changed, or the summary is missing rows |
| Ranking figures | `python scripts/plot_results.py` | no | after the summary has four keys |

Skip the two GPU cells if Drive already has four keys in `pilot_summary_default.json` (this checkout restored them from the frozen `*_default.json` files). Do **not** run both GPU cells unless you want a learned-only exam and then a full rescore.

Needs `results/checkpoints/learned_policy.pt` for the GPU cells. Train first if it is missing:

```bash
python scripts/train_policy.py
```

See `docs/HOW_TO_RUN.md` §4.6.

In [ ]:
# Merge unit test + show whether the default summary already has four policies
from pathlib import Path
import json

!python tests/test_pilot_summary_merge.py

p = Path("results/metrics/pilot_summary_default.json")
ckpt = Path("results/checkpoints/learned_policy.pt")
print("learned_policy.pt exists:", ckpt.exists(), "->", ckpt)
if not p.exists():
    print("pilot_summary_default.json missing — run the four-policy cell")
else:
    keys = list(json.loads(p.read_text())["results"])
    print("pilot_summary_default.json keys:", keys)
    want = ["naive_rag", "rule_based", "max_tools", "learned"]
    if keys == want:
        print("Four-policy table is already complete. Skip the GPU cells unless the reward or slice changed.")
    else:
        print("Missing", [k for k in want if k not in keys], "— run the four-policy cell (needs learned_policy.pt).")


In [ ]:
# Optional GPU — learned-only exam on the default preset.
# Merges the learned row into pilot_summary_default.json; does not drop rule / max.
# Skip if the previous cell said the four-policy table is already complete.
!python scripts/run_pilot.py --policies learned


In [ ]:
# Optional GPU — four-policy rebuild (naive / rule / max / learned).
# Use after a reward or slice change so every row shares the new setup.
# Needs results/checkpoints/learned_policy.pt (skipped learned if missing).
!python scripts/run_pilot.py --run-env-check


In [ ]:
# Refresh ranking figures from pilot_summary_default.json (no GPU)
!python scripts/plot_results.py


In [ ]:
# ##Git Pull
!git log -3
!git status

In [31]:
##Git Push
from google.colab import userdata

TOKEN = userdata.get("GITHUB_TOKEN")
!git config --global user.name "jitu"
!git config --global user.email "jitu.learn@gmail.com"

!git remote set-url origin https://Smartcobra:{TOKEN}@github.com/Smartcobra/ca-rl-arag.git
##Push
!git status
!git add \
results/checkpoints/learned_policy_frontier_lambda0.pt \
results/checkpoints/learned_policy_frontier_lambda0_best.pt \
results/checkpoints/learned_policy_frontier_lambda20.pt \
results/checkpoints/learned_policy_frontier_lambda20_best.pt \
results/checkpoints/learned_policy_frontier_act02.pt \
results/checkpoints/learned_policy_frontier_act02_best.pt \
results/figs/frontier_em_usd.png \
results/metrics/frontier_sweep_table.json \
results/metrics/learned_frontier_lambda0.json \
results/metrics/learned_frontier_lambda20.json \
results/metrics/learned_frontier_act02.json \
results/metrics/pilot_summary_frontier_lambda0.json \
results/metrics/pilot_summary_frontier_lambda20.json \
results/metrics/pilot_summary_frontier_act02.json \
results/metrics/train_policy_curve_frontier_lambda0.json \
results/metrics/train_policy_curve_frontier_lambda20.json \
results/metrics/train_policy_curve_frontier_act02.json \
results/trajectories/learned_frontier_lambda0.jsonl \
results/trajectories/learned_frontier_lambda20.jsonl \
results/trajectories/learned_frontier_act02.jsonl \
notebooks/Frontier_Cost_Pressure_CA_RL_ARAG.ipynb

!git commit -m "Colab-Execution: learned cost-pressure frontier (lambda 0/20/80, 40 epochs, greedy eval)"
!git push -u origin feature_baseline_v2

On branch feature_baseline_v2
Your branch is up to date with 'origin/feature_baseline_v2'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	results/checkpoints/learned_policy_frontier_act0.pt
	results/checkpoints/learned_policy_frontier_act005.pt
	results/checkpoints/learned_policy_frontier_act005_best.pt
	results/checkpoints/learned_policy_frontier_act02.pt
	results/checkpoints/learned_policy_frontier_act02_best.pt
	results/checkpoints/learned_policy_frontier_act0_best.pt
	results/figs/frontier_em_usd.png
	results/metrics/frontier_sweep_table.json
	results/metrics/learned_frontier_act0.json
	results/metrics/learned_frontier_act005.json
	results/metrics/learned_frontier_act02.json
	results/metrics/pilot_summary_frontier_act0.json
	results/metrics/pilot_summary_frontier_act005.json
	results/metrics/pilot_summary_frontier_act02.json
	results/metrics/train_policy_curve_frontier_act0.json
	results/metrics/train_policy_curve_frontier_act005.json
	results/m